pkg/kubecost/allocation.go

```diff
@@ -2131,34 +2131,179 @@ func (asr *AllocationSetRange) NewAccumulation() (*AllocationSet, error) {
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error accumulating result: %s", err)
+		}
+		toAccumulate = NewAllocationSetRange(asAccumulated)
+
+		if hour == 23 || i == len(asr.Allocations)-1 {
+			if length := len(toAccumulate.Allocations); length != 1 {
+				return nil, fmt.Errorf("failed accumulation, detected %d sets instead of 1", length)
+			}
+			result.Append(toAccumulate.Allocations[0])
+			toAccumulate = nil
 		}
+	}
+	return result, nil
+}
 
-		if allocSet != nil {
+func (asr *AllocationSetRange) accumulateByMonth(clone bool) (*AllocationSetRange, error) {
+	var toAccumulate *AllocationSetRange
+	result := NewAllocationSetRange()
+	for i, as := range asr.Allocations {
+		if as.Window.Duration() != time.Hour*24 {
+			return nil, fmt.Errorf("window duration must equal 24 hours; got:%s", as.Window.Duration())
+		}
 
-			// check if end of asr to sum the final set
-			// If total asr accumulated sum <= resolution return 1 accumulated set
-			if allocSet.Window.Duration() >= resolution || i == len(asr.Allocations)-1 {
-				allocSetRange.Allocations = append(allocSetRange.Allocations, allocSet)
-				allocSet = NewAllocationSet(time.Time{}, time.Time{})
+		if clone {
+			as = as.Clone()
+		}
+
+		_, month, _ := as.Window.Start().Date()
+		_, nextDayMonth, _ := as.Window.Start().Add(time.Hour * 24).Date()
+
+		if toAccumulate == nil {
+			toAccumulate = NewAllocationSetRange()
+		}
+
+		toAccumulate.Append(as)
+		asAccumulated, err := toAccumulate.accumulate()
+		if err != nil {
+			return nil, fmt.Errorf("error accumulating result: %s", err)
+		}
+		toAccumulate = NewAllocationSetRange(asAccumulated)
+
+		// either the month has ended, or there are no more allocation sets
+		if month != nextDayMonth || i == len(asr.Allocations)-1 {
+			if length := len(toAccumulate.Allocations); length != 1 {
+				return nil, fmt.Errorf("failed accumulation, detected %d sets instead of 1", length)
 			}
+			result.Append(toAccumulate.Allocations[0])
+			toAccumulate = nil
 		}
 	}
+	return result, nil
+}
+
+func (asr *AllocationSetRange) accumulateByWeek(clone bool) (*AllocationSetRange, error) {
+	var toAccumulate *AllocationSetRange
+	result := NewAllocationSetRange()
+	for i, as := range asr.Allocations {
+		if as.Window.Duration() != time.Hour*24 {
+			return nil, fmt.Errorf("window duration must equal 24 hours; got:%s", as.Window.Duration())
+		}
 
-	return allocSetRange, nil
+		if clone {
+			as = as.Clone()
+		}
+
+		dayOfWeek := as.Window.Start().Weekday()
+
+		if toAccumulate == nil {
+			toAccumulate = NewAllocationSetRange()
+		}
+
+		toAccumulate.Append(as)
+		asAccumulated, err := toAccumulate.accumulate()
+		if err != nil {
+			return nil, fmt.Errorf("error accumulating result: %s", err)
+		}
+		toAccumulate = NewAllocationSetRange(asAccumulated)
+
+		// current assumption is the week always ends on Saturday, or there are no more allocation sets
+		if dayOfWeek == time.Saturday || i == len(asr.Allocations)-1 {
```

Comment:
That is an accurate understanding. Does this scenario happen?

My main goal was to avoid difficult to understand date parsing code (especially when we have to come back and add a configurable start of the week day), but if we need it, we need it.

Reply:
Frankly, I'm not sure. @nikovacevic if Kubecost is down for >= 1 day, is it possible to query and get an ASR with entire AllocatioSet(s) missing?

Reply:
No, there shouldn't be any missing days. In that case, there would be an _empty_ day, or days, but that AllocationSet should still exist with a window.

Reply:
I've seen something like my theoretical situation _serialized_ (JSON) as `ASR{ {}, {}, {}, {filled...} ...}`. It seems that `nil` ASes get turned into `{}` by our custom marshaler: https://github.com/opencost/opencost/blob/675a6af3f8ccb2e289fdc61d342821effd7c06cb/pkg/kubecost/allocation_json.go#L169-L175

That indicates that we could have a `nil` AS somewhere in an ASR, taking my earlier example:
```
ASR{
    AS{ window: thursday }
    AS{ window: friday }
    nil
    AS{ window: sunday }
}
```

I'll caveat this with the fact that I've never seen the `{}` show up in the _middle_ of an ASR, but maybe its possible?

Reply:
I'm still a little concerned about this, but I think it is sufficiently edge that we can proceed for now. If we do run into the hypothetical bug, it should at least show up in the resulting Window, meaning it will be debuggable.

Reply:
I don't think concern is altogether unwarranted! Haha. You're right about the JSON side of things, but I empty AllocationSets that are _not_ null also encode that way. And because we never save `nil` sets, but we do save empty ones, I'd be very surprised if we ever had a nil.

(Feel free to keep this receipt in the case that I am wrong :wink: )